pkg/efivarfs/varfs.go

```diff
@@ -0,0 +1,243 @@
+	Delete(name string, guid guid.GUID) error
+	List() ([]VariableDescriptor, error)
+}
+
+type varfsBackend struct{}
+
+// Get reads the contents of an efivar if it exists and has the necessary permission
+func (v varfsBackend) Get(name string, guid guid.GUID) (VariableAttributes, []byte, error) {
+	path := filepath.Join(EfiVarFs, fmt.Sprintf("%s-%s", name, guid))
+	f, err := openVarfsFile(path, os.O_RDONLY, 0)
+	switch {
+	case os.IsNotExist(err):
+		return 0, nil, ErrVarNotExist
+	case os.IsPermission(err):
+		return 0, nil, ErrVarPermission
+	case err != nil:
+		return 0, nil, err
+	}
+	defer f.Close()
+
+	var attrs VariableAttributes
+	err = binary.Read(f, binary.LittleEndian, &attrs)
+	if err != nil {
+		if err == io.EOF {
+			return 0, nil, ErrVarNotExist
+		}
+		return 0, nil, err
+	}
+
+	data, err := io.ReadAll(f)
+	if err != nil {
+		return 0, nil, err
+	}
+	return attrs, data, nil
+}
+
+// Set modifies a given efivar with the provided contents
+func (v varfsBackend) Set(name string, guid guid.GUID, attrs VariableAttributes, data []byte) (bool, error) {
+	path := filepath.Join(EfiVarFs, fmt.Sprintf("%s-%s", name, guid))
+	flags := os.O_WRONLY | os.O_CREATE
+	if attrs&AttributeAppendWrite != 0 {
+		flags |= os.O_APPEND
+	}
+
+	read, err := openVarfsFile(path, os.O_RDONLY, 0)
+	switch {
+	case os.IsNotExist(err):
+	case os.IsPermission(err):
+		return false, ErrVarPermission
+	case err != nil:
+		return false, err
+	default:
+		defer read.Close()
+
+		restoreImmutable, err := makeVarFileMutable(read)
+		switch {
+		case os.IsPermission(err):
+			return false, ErrVarPermission
+		case err != nil:
+			return false, err
+		}
+		defer restoreImmutable()
+	}
+
+	write, err := openVarfsFile(path, flags, 0644)
+	switch {
+	case os.IsPermission(err):
+		pe, ok := err.(*os.PathError)
+		if !ok {
+			return false, err
+		}
+		if pe.Err == syscall.EACCES {
+			// open will fail with EACCES if we lack the privileges
+			// to write to the file or the parent directory in the
+			// case where we need to create a new file. Don't retry
+			// in this case.
+			return false, ErrVarPermission
+		}
+
+		// open will fail with EPERM if the file exists but we can't
+		// write to it because it is immutable. This might happen as a
+		// result of a race with another process that might have been
+		// writing to the variable or may have deleted and recreated
+		// it, making the underlying inode immutable again. Retry in
+		// this case.
+		return true, ErrVarPermission
+	case err != nil:
+		return false, err
+	}
+	defer write.Close()
+
+	var buf bytes.Buffer
+	err = binary.Write(&buf, binary.LittleEndian, attrs)
+	if err != nil {
+		return false, err
+	}
+	for len(data)%8 != 0 {
+		data = append(data, 0)
+	}
+	size, err := buf.Write(data)
+	if err != nil {
+		return false, err
+	}
+	if (size-4)%8 == 0 {
+		return false, fmt.Errorf("data misaligned")
+	}
+	_, err = buf.WriteTo(write)
```

Comment:
it's common in this case to to
if _, err := but.WriteTo(...); err != nil

Reply:
It would be the only place in that package where I'd then use that syntax and the error check is there so I'd rather prefer to keep the syntax uniform across a package